Full refactoring and improvement of the project

cernyd opened the pull request:
I thought this project was an interesting idea and wanted to improve it, I ended up pretty much rewriting most of the code structure.

## Findings
1) The overuse of classes created IMHO more obstacles than actual value, a functional structure is better suited for a command line utility that only uses the results once
    - Instead of instances, more simple data types are being used now
2) Instead of redeveloping your own argument parser, I recommend using the very powerful `argparse` python already offers, it leads to easier addition of new args and avoids pitfalls, the help menu is generated automatically so there is no need to update it manually
4) Regex is a good tool but it is better to avoid it when parsing HTML, I propose using the BS4 library that can safely parse HTML and easily traverse all elements on the page (in case you wanted to add some more information gathering in the future)

## Improvements
1) Rewrote the table rendering logic to simplify it
2) Improved overall code clarity
3) Added proper documentation
4) Rewrote block and room validation logic, corresponding helpful error messages are tailored to up to date data in `const.py`
5) I changed some variable and function names to more descriptive ones and wrote a few utilites.
6) The source is now a lot shorted with less files, the less code, the less bugs can occur.
7) The female name detection can now easily be enhanced by using blacklists and whitelists instead of changing code logic.
8) In case you want to know something, feel free to ask

## Summary
I hope you will acknowledge my effort, there might be some bugs/inconsistencies within the new code since I first read the source yesterday but all of that can be resolved in this PR if you add feedback, thanks!

pep8speaks commented:
Hello @cernyd! Thanks for opening this PR.

* In the file [`const.py`](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/const.py):

> [Line 16:101](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/const.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 100 characters)
> [Line 18:101](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/const.py#L18): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)

* In the file [`finder.py`](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/finder.py):

> [Line 26:1](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/finder.py#L26): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`knbh.py`](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/knbh.py):

> [Line 33:20](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/knbh.py#L33): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 43:19](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/knbh.py#L43): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 104:1](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/knbh.py#L104): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`utils.py`](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py):

> [Line 32:5](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py#L32): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 33:32](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py#L33): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 33:43](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py#L33): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 52:1](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py#L52): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 62:1](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py#L62): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 70:1](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py#L70): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 90:101](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py#L90): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 103:101](https://github.com/europ/KNBH/blob/8aca09173e02754a98a1dcbcb454995af29213a7/utils.py#L103): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 100 characters)

europ commented:
Firstly, thank you for your contribution.

The review will be delayed, I do not have currently time for that. I and @tammar96 will review your PR, hopefully as soon as possible (less than 3 weeks). As I did a super quick review, I have an advice for you.

> #### Improvements
> 1. Rewrote the table rendering logic to simplify it

You did not reach that. By removing my `"{}".format(var)` and rewriting it to string join the problem is still unresolved. Please, can you integrate https://pypi.org/project/tabulate/ to make it simple and logic? Here is an inspirating topic for you https://stackoverflow.com/a/26937531. To make python code simple and clean, you must use libraries such this, which are designed for this specific purpose.

davidbolvansky commented on const.py:
```diff
@@ -1,45 +1,71 @@
+
+# Can be easily edited if need be, the program adjusts automatically
+
+DORM_ROOMCOUNTS = {
```
Can we specify which room number is first?

Manesovy koleje - we have no room eg. 300, we start from 301.

europ commented on const.py:
```diff
@@ -1,45 +1,71 @@
+}
+
+LAST_NAME = {
+    "whitelist": ("a", "á", "ová", "ova"),
```
`ová` or `ova` are useless, it's enough to match `á` or `a`

cernyd replied:
From what I have seen `a` created a lot of false positives, there are many male surnames that end with `a`, maybe it would be better only to keep `ová`?

europ replied:
Please, provide some examples that confirm this statement - man surnames end with a/á.

davidbolvansky commented on README.md:
```diff
@@ -26,41 +26,34 @@ Define woman as a person with surname ending with `á` or firstname ending with
-    - `./main.py -b=B05 -f=2`
-    - `python3.6 main.py -b=B05 -f=2`
+    - `./knbh.py -b B05 -f 2`
+    - `python3.6 knbh.py -b B05 -f 2`
```
Yeah, agree

europ replied:
The option modifies the output as follows:
1.) only empty rooms or partially occupied rooms, which indicates that there is a free space for somebody;
2.) all rooms including empty or partially empty.

cernyd replied:
Definition of empty room = room with no results that is also in the valid room range? @europ

europ replied:
Exclude non-existing rooms. The room represents a physically existing and valid room of the dormitory.

Please, add two output modifying parameters:
1.) `--include-empty` / `-e`
2.) `--empty-only `/ `-E`
whose behavior is as follows:
1.) list everything (literally), empty, partially empty, fully occupied from existing rooms
2.) list only empty and partially empty rooms from existing rooms - rooms, where is still a possibility of accommodation

cernyd commented:
Added an argument to cache query results so we don't have to bother the server each time when developing knbh, also added option to export to excel and csv (also WIP). Tabular is almost ready too. @europ @davidbolvansky @tammar96

cernyd commented:
@europ I would not advise merging it just yet, there are still some minor things I need to finish up. I'll tell you when ready.

europ commented:
Do not worry, firstly we have to check it in detail. I do not know how should @tammar96 approve this so quickly. It's a neglected review.

europ commented on Makefile:
```diff
@@ -0,0 +1,6 @@
+help:
+	python3 knbh.py --help
+install:
+	pip3 install -r requirements.txt
```
Why is this necessary? How to install requirements should be added to Help section in README.md. This is not a C code you need a Makefile...

cernyd replied:
Not necessary but standard practice in many non-c projects. I mainly wrote it for quick debugging without having to search for commands in bash history or retyping them. @europ

cernyd replied:
make install in particular might not be needed, it is just my personal preference.

europ replied:
I would like to request you to remove this, there is no need of this. It's unnecessary. @davidbolvansky what do you think?

davidbolvansky replied:
Maybe rename to make install_deps?

cernyd replied:
I’ll remove it. @europ

europ replied:
Makefile was not removed.

The `README.md` file should contain all the relevant information for a basic setup and launch such as 1) prerequisites 2) dependencies (installation) 3) how to run (usually some test e.g. run --help).

europ replied:
:warning: @cernyd this is not resolved

davidbolvansky commented:
What is status of this PR?

cernyd commented:
Waiting for a final review as @europ said. I currently don't have time to work on this, could continue after exams in january. @davidbolvansky

europ commented:
@cernyd I am sorry, currently, I am not available. I will review this pull request on 10th January 2020.

europ commented on knbh.py:
```diff
@@ -0,0 +1,204 @@
+        # Autogenerated rooms don't need to be validated
+    else:
+        print("Supply either a room number or a floor number to search!")
+        exit(-1)
```
The used number should be from the range of <0; 255>

If you try `python -c "import sys; sys.exit(-1)"` it will return 255. Use positive nonzero numbers.

https://stackoverflow.com/questions/44893807/i-want-to-know-what-exactly-sys-exit-1-returns-in-python

europ commented on knbh.py:
```diff
@@ -0,0 +1,204 @@
+        action = export_datatype if export_datatype else "print"
+        process_data(query_result, action)
+    else:
+        print("No results found...")
```
You are using print while the stderr/stdout/exit commands are wrapped.

Please, you decide, if you want to use those (in system.py) or not. What's better for you?

If not please remove the system file as well because it will be useless. But the required function such as signal function should be moved to the file where they are used.

europ replied:
:warning: @cernyd this is not resolved

europ commented on utils.py:
```diff
@@ -0,0 +1,175 @@
+        filename = "table_%s_floor_%s.xlsx" % (data["block"], data["floor"])
+        df.to_excel(filename)
+        print("Results saved to file '%s'" % filename)
+    # print(df.loc[df['female']])
```
Fix this comment.

europ commented on knbh.py:
```diff
@@ -0,0 +1,204 @@
+
+    args = vars(PARSER.parse_args())
+
+    if len(args) == 0:  # TODO
```
Fix this TODO.

europ commented on knbh.py:
```diff
@@ -0,0 +1,204 @@
+
+    if cached:
+        try:
+            with open("cache.json", "r") as file:
```
Add this `cache.json` file to `.gitignore` file otherwise git will track it.

cernyd commented:
Thank's for the review @europ, I'll resolve the issues when I'll have time for it.

davidbolvansky commented:
I would prefer to retain original “TUI” logic, especially the girls marked with red color instead of the “female” column. Since we specifiy the block number too, I dont see why we need a new column for it. Also top left column with counter seems useless for me - if you want, you can add this info to the title - eg. B05 2. floor (78 people) or so..

europ commented:
As mentioned [above](https://github.com/europ/KNBH/pull/20#issuecomment-569931523), please, keep the original functionality (the result was printed to output as a colored table).

cernyd commented:
@europ is there any particular reason why python 3.6 is specifically required? (as opposed to simply python3)

europ commented on README.md:
```diff
@@ -26,41 +26,34 @@ Define woman as a person with surname ending with `á` or firstname ending with
-    - `./main.py -h`
-    - `python3.6 main.py -h`
+    - `./knbh.py -h`
+    - `python3.6 knbh.py -h`
```
> is there any particular reason why python 3.6 is specifically required? (as opposed to simply python3)

we use python3 (see knbh.py 1st line) ... can you please fix this from python3.6 to python3?

davidbolvansky commented:
Can you please resolve few remaining issues?

cernyd commented:
@davidbolvansky hi, sorry for not responding, I'll finish the changes as soon as possible.

cernyd commented:
@europ @davidbolvansky fixed the remaining issues, the last problem is I could not find any partially empty and empty rooms to test my code on. Otherwise you can test for yourself, everything should be working as intended. I also rewrote the argument parsing again, this time with the [click library](https://click.palletsprojects.com/en/7.x/). excel and csv exports now working, the tables are formatted as before - the table width is now determined by the maximum name length, I think that's more flexible than a static width but you decide whether to keep it.

cernyd commented:
Also, could you add me to code owners since I rewrote many of the files completely? @europ

davidbolvansky commented:
This does not work: python3 knbh.py --block d02 --floor 3
This works: python3 knbh.py --block D02 --floor 3

Please fix it.
-----------------------
-f=X or -b=Y options work just fine with current master branch. Your branch:
xbolva00@xbolva00-G551JW:~/KNBH/KNBH$ python3 knbh.py -b=d02 -f=3
Usage: knbh.py [OPTIONS]
Try "knbh.py --help" for help.

Error: Invalid value for "--floor" / "-f": =3 is not a valid integer

This is broken, a fix is needed. Please ensure that "-opt=VAL" format still works for every option.
-----------------------
Anyway, new terminal UI is weird..

![image](https://user-images.githubusercontent.com/22410393/74328200-cccdfa80-4d8d-11ea-8a7e-957e77d50f78.png)

You can add option to "generate web page" but please, by default, preserve original view like:

![image](https://user-images.githubusercontent.com/22410393/74328318-eec77d00-4d8d-11ea-9899-0479a6e4daea.png)
-----------------------
Also, python3 knbh.py --block B05 -f 2 there is a crash:
![image](https://user-images.githubusercontent.com/22410393/74328501-52ea4100-4d8e-11ea-9169-d3d146349410.png)

cernyd commented:
Hi, sorry for not responding once again, I'll resolve this ASAP.

cernyd commented:
@davidbolvansky I resolved the issues you described above, the HTML was just a debug print I forgot to remove. The `-opt=VAL` works but only as `-opt VAL` without the `=`, is that an issue?

davidbolvansky commented:
Thanks !

Yeah, I would like to support -opt=VAL too. What do you think @europ ? @tammar96 ?

I will try your branch later today to give you last comments (if any)

europ commented:
If the arguments are in `-o=val`/`--opt=val` or `-o val`/`--opt val` it's alright. It does not matter after all. But the purpose of this project should be preserved.

davidbolvansky commented:
Female detector is totally broken. Try: python3 knbh.py -bD02 -f3 or python3 knbh.py -bB05 -f2 

Either revert to previous female detection logic or fix it.


After you resolve this issue, LGTM.

cernyd commented:
@davidbolvansky the issue with name detection should now be resolved, try for yourself.

davidbolvansky commented:
Thanks for fast fix, looks good.

davidbolvansky commented:
> Also, could you add me to code owners since I rewrote many of the files completely? @europ

Please @europ add him to contributors :)

davidbolvansky commented:
Please, rebase your branch and squash all your partial commits to one.

cernyd commented:
@davidbolvansky done.

europ commented on knbh.py:
```diff
@@ -0,0 +1,118 @@
+            with open("cache.json", "r") as file:
+                query_result = json.loads(file.read())
+        except Exception:
+            pass
```
what? if you cannot open the file, why you do not print any error message?

europ commented on knbh.py:
```diff
@@ -0,0 +1,118 @@
+            query_result[block]["rooms"].append(room_data)
+
+        # Cache last result in case it is needed
+        with open("cache.json", "w") as file:
```
except EnvironmentError exception

europ commented on utils.py:
```diff
@@ -0,0 +1,231 @@
+    :param room_number: {str} room number to be checked
+    """
+
+    try:  # If room number can't be converted to a number, it can't be a number
```
remove this useless comment

europ commented on knbh.py:
```diff
@@ -0,0 +1,118 @@
+        print_block_map()
+        exit(1)
+    else:
+        block = block.upper()
```
```python
print_block_map()
exit(1)
```
vs.
```python
print_block_map()
return
```
use return or exit, do not use both

europ commented on utils.py:
```diff
@@ -0,0 +1,231 @@
+
+    table += divider
+
+    return table
```
* https://github.com/europ/KNBH/pull/20#issuecomment-542696322
* https://pypi.org/project/tabulate/

where is it?

europ commented on README.md:
```diff
@@ -6,61 +6,56 @@
+     --empty-only          only list rooms that are not full
+     -c, --cached          use cached query data instead of scraping it
+     --export [excel|csv]  export to one of these supported formats: excel, csv
+     --help                Show this message and exit.
```
add `-h` too

cernyd commented:
Hi @europ, I fixed the issues you mentioned. I also implemented tabular. However, using tabular creates multiple issues that are not easily solveable:

1. Tabular doesn't easily support grouping multiple rows "by room", so I had to put multiple names into one column so that tabular would separate by room
2. Tabular breaks if I use newlines and colored text at the same time, I had to use `, ` to join instead of `\n`
<img width="275" alt="obrazek" src="https://user-images.githubusercontent.com/18537381/74873610-d04e1c80-535f-11ea-803f-9fc990db2d62.png">

This is the only way I can present the data without breaking the table format right now:
<img width="389" alt="obrazek" src="https://user-images.githubusercontent.com/18537381/74873697-fb387080-535f-11ea-921d-7e59f8e9579a.png">

If you don't like it I can revert the tabular implementation, but I won't be trying to make tabular work with our tables.

I think the `termcolor` is outside of scope of this merge request.

Please decide whether to merge this or not as I can't keep rewriting it, it has already taken far too much of my time, I hope you understand.

davidbolvansky commented:
New UI is fine for me.

LGTM.

davidbolvansky commented:
ping @europ

europ commented:
* rebase this branch
* test it :bangbang: 
* make it sure that it will pass the following command: `pycodestyle --config=./.conf/pycodestyle.cfg . && flake8 --config=./.conf/flake8.cfg . && mypy --ignore-missing-imports . && mdl --style ./.conf/markdownlint.rb .`

If you do this, the branch will be merged by @davidbolvansky.
